### DISCUSSION TITLE
How to convert SQL to DMMF?

### DISCUSSION THREAD

**Discussion by mohammed-bahumaish** [prisma/prisma#18286] [Q&A] [answered]
Hi everyone, I'm working on an opensource project and I need to implement the following:
- a function that takes SQL table definitions as a string (as in migrations folder of Prisma) and returns the DMMF(or Prisma schema).

is there an internal Prisma package that does that ? I would be grateful if you can help me solve this.

example:

## input:

        -- CreateEnum
        CREATE TYPE "Role" AS ENUM ('USER', 'ADMIN');
  
        -- CreateTable
        CREATE TABLE "User" (
            "id" SERIAL NOT NULL,
            "email" TEXT NOT NULL,
            "name" TEXT,
        
            CONSTRAINT "User_pkey" PRIMARY KEY ("id")
        );
        
        -- CreateTable
        CREATE TABLE "Post" (
            "id" SERIAL NOT NULL,
            "createdAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
            "updatedAt" TIMESTAMP(3) NOT NULL,
            "title" TEXT NOT NULL,
            "content" TEXT,
            "published" BOOLEAN NOT NULL DEFAULT false,
            "viewCount" INTEGER NOT NULL DEFAULT 0,
            "authorId" INTEGER,
        
            CONSTRAINT "Post_pkey" PRIMARY KEY ("id")
        );
        
        -- CreateIndex
        CREATE UNIQUE INDEX "User_email_key" ON "User"("email");
        
        -- AddForeignKey
        ALTER TABLE "Post" ADD CONSTRAINT "Post_authorId_fkey" FOREIGN KEY ("authorId") REFERENCES "User"("id") ON DELETE SET NULL ON UPDATE CASCADE;

## output

      model User {
        id    Int     @id @default(autoincrement())
        email String  @unique
        name  String?
        posts Post[]
      }
      
      model Post {
        id        Int      @id @default(autoincrement())
        createdAt DateTime @default(now())
        updatedAt DateTime @updatedAt
        title     String
        content   String?
        published Boolean  @default(false)
        viewCount Int      @default(0)
        author    User?    @relation(fields: [authorId], references: [id])
        authorId  Int?
      }
                
      enum Role {
        USER
        ADMIN
      }

**Comment by janpio** [comment, marked as answer]
Prisma does not support that as far as I know. You would need to run the SQL on a database first, and then use `db pull` to get a schema representation.

What is your full use case?

**Reply by mohammed-bahumaish** [reply]
I'm doing it that way, but it is slow and I'm looking for a faster way that does not include a live database.
this is the project I'm working on
 [Prisma Editor](https://prisma-quick-production.up.railway.app/). Prisma editor is a powerful tool to visualize and edit Prisma Schema.
I'm still developing it. and experimenting with AI for schema generating. I might remove the SQL part of it, or write a parser that converts SQL to Prisma. 

anyways thanks for your answer @Janpio.

**Reply by janpio** [reply]
Sweet!

I don't think we have any capability to turn a SQL string into a Prisma schema without the in between step of running that SQL on a database. I am not even sure if that is technically possible.

**Reply by ludralph** [reply]
@mohammed-bahumaish you might want to take a look at this [SQL to Prisma Schema generator project](https://github.com/Devzstudio/SQL-to-Prisma-Schema-Generator)
